### PR TITLE
GitHub Actions for macOS

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

**Comment by reuben** [comment]
Could you push again to make this run?

File: .github/actions/check_artifact_exists/main.js
Comment (reuben) [inline .github/actions/check_artifact_exists/main.js:121]: What is this adding that simply letting it fail without a global try-catch doesn't do?

File: .github/actions/check_artifact_exists/package.json
Comment (reuben) [inline .github/actions/check_artifact_exists/package.json:10]: When you move the action documentation into this folder, can you reference the official docs or write some words yourself on how to handle changes to dependencies? I assume it'll be something like "change package.json, `npm install`, test, commit `node_modules` folder and `package-lock.json` file" but it's good to have it documented. If this is covered in official GitHub docs then just a link would be enough.

File: .github/workflows/build.yml
Comment (reuben) [inline .github/workflows/build.yml:133]: nit (throughout the PR): GitHub style seems to be to use-dashes-in-names rather than underscores. In this PR they're mixed. Job-names use dashes, action_inputs use underscores, and in this matrix you have both. I think it's better to match GitHub's style and use-dashes-everywhere.

**Review by reuben — APPROVED** [review]
Some comments but generally looks very good 🚀

**Comment by lissyx** [comment]
> Could you push again to make this run?

You mean to have it running against `mozilla/DeepSpeech` repo? I think it would run until we merge, but you can check on my fork :)

File: .github/workflows/build.yml
Comment (lissyx) [reply .github/workflows/build.yml:133]: Good catch. FTR for now there's a constraint in the `get_cache_key` job that I rely on `_` to separate between the job's name of a submodule and the platform, to make computation of the key smoother and with less parameters. I will have to ensure I properly document that in the action itself.

File: .github/actions/check_artifact_exists/package.json
Comment (lissyx) [reply .github/actions/check_artifact_exists/package.json:10]: Sure. One thing I want to test is if it's better to NOT commit `node_modules` and rely on an extra `npm install` step, but I fear it might make the action more fragile (remember all the infras issues because of npm).

File: .github/actions/check_artifact_exists/main.js
Comment (lissyx) [reply .github/actions/check_artifact_exists/main.js:121]: It might have been me trying to be extra cautious / having proper debug output

File: .github/actions/build-tensorflow/action.yml
Comment (lissyx) [reply .github/actions/build-tensorflow/action.yml:7]: I'll drop the `default` if you agree and stick to `required: true`

File: .github/actions/README.md
Comment (lissyx) [reply .github/actions/README.md:35]: Indeed. Do you think we should also document the workflow ?

File: .github/actions/README.md
Comment (lissyx) [reply .github/actions/README.md:23]: Correct, I had it renamed in the meantime. Also a very good reason why this should live within the action as much as possible.

File: .github/actions/README.md
Comment (reuben) [reply .github/actions/README.md:35]: I think giving names to not so transparent steps such as "host-build" might be enough. There's nothing in the workflow currently that seems particularly obscure.

File: .github/actions/build-tensorflow/action.yml
Comment (reuben) [reply .github/actions/build-tensorflow/action.yml:7]: Yep sounds good.

File: .github/actions/check_artifact_exists/package.json
Comment (reuben) [reply .github/actions/check_artifact_exists/package.json:10]: Yeah seems like fewer live `npm` calls are best from our experience. Maybe we can move to the GitHub recommended step of checking in the compiled `dist/index.js` file instead of node_modules. But I think we don't need to block on that (or even do that work ourselves).

**Comment by reuben** [comment]
> > Could you push again to make this run?
> 
> You mean to have it running against `mozilla/DeepSpeech` repo? I think it would run until we merge, but you can check on my fork :)

Yes. I enabled Actions on this repo. I think instead of working on a giant "GitHub Actions" PR we should rename this to "macOS GitHub Actions" and land ASAP and keep working in separate PRs. And for that we need to get it running here.

File: .github/actions/README.md
Comment (lissyx) [reply .github/actions/README.md:35]: > I think giving names to not so transparent steps such as "host-build" might be enough. There's nothing in the workflow currently that seems particularly obscure.

Right, what suggestion for `host-build` ? It was pretty transparent in my mind ;)

**Comment by lissyx** [comment]
> > > Could you push again to make this run?
> > 
> > 
> > You mean to have it running against `mozilla/DeepSpeech` repo? I think it would run until we merge, but you can check on my fork :)
> 
> Yes. I enabled Actions on this repo. I think instead of working on a giant "GitHub Actions" PR we should rename this to "macOS GitHub Actions" and land ASAP and keep working in separate PRs. And for that we need to get it running here.

Sure, that was my goal as well

File: .github/actions/check_artifact_exists/package.json
Comment (lissyx) [reply .github/actions/check_artifact_exists/package.json:10]: I need to re-verify that

File: .github/actions/README.md
Comment (reuben) [reply .github/actions/README.md:35]: Btw I'm talking about the job name, not the action name. So I imagined something like "Build libdeepspeech.so and deepspeech binary"

Comment (lissyx) [reply .github/actions/README.md]: > Btw I'm talking about the job name, not the action name. So I imagined something like "Build libdeepspeech.so and deepspeech binary"

The job name is defined in the workflow and we are not completely free on the string.

**Comment by lissyx** [comment]
Manual testing of the TF / TFLite NodeJS (NodeVS v15):
```
ds-tc-worker:test-github-actions ds-worker-admin$ PATH=$(pwd)/node-v15.12.0-darwin-x64/bin/:$PATH ./tflite/node_modules/.bin/deepspeech --model deepspeech-0.9.3-models.tflite --scorer deepspeech-0.9.3-models.scorer --audio audio/4507-16021-0012.wav --extended --stream
Loading model from file deepspeech-0.9.3-models.tflite
TensorFlow: v2.3.0-6-g23ad988fcd
DeepSpeech: v0.10.0-alpha.3-57-g0b30a4d8
Loaded model in 0.005774s.
Loading scorer from file deepspeech-0.9.3-models.scorer
Loaded scorer in 0.0004919s.
intermediate:
intermediate:
intermediate:
intermediate:
intermediate: why
intermediate: why should i
intermediate: why should one
intermediate: why should one
intermediate: why should one halt to
intermediate: why should one halt on the
intermediate: why should one halt on the
why should one halt on the way
ds-tc-worker:test-github-actions ds-worker-admin$ PATH=$(pwd)/node-v15.12.0-darwin-x64/bin/:$PATH ./tf/node_modules/.bin/deepspeech --model deepspeech-0.9.3-models.pbmm --scorer deepspeech-0.9.3-models.scorer --audio audio/4507-16021-0012.wav --extended --stream
Loading model from file deepspeech-0.9.3-models.pbmm
TensorFlow: v2.3.0-6-g23ad988fcd
DeepSpeech: v0.10.0-alpha.3-57-g0b30a4d8
2021-03-24 15:00:28.072326: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Loaded model in 0.01582s.
Loading scorer from file deepspeech-0.9.3-models.scorer
Loaded scorer in 0.0004769s.
intermediate:
intermediate:
intermediate:
intermediate:
intermediate: why s
intermediate: why should i
intermediate: why should one
intermediate: why should one
intermediate: why should one halt
intermediate: why should one halt on the
intermediate: why should one halt on the
why should one halt on the way
ds-tc-worker:test-github-actions ds-worker-admin$
```

**Comment by lissyx** [comment]
Manual testing of the TF / TFLite Python (Python 3.9 Homebrew):
```
ds-tc-worker:test-github-actions ds-worker-admin$ deepspeech --model deepspeech-0.9.3-models.pbmm --scorer deepspeech-0.9.3-models.scorer --audio audio/4507-16021-0012.wav --extended
Loading model from file deepspeech-0.9.3-models.pbmm
TensorFlow: v2.3.0-6-g23ad988fcd
DeepSpeech: v0.10.0-alpha.3-57-g0b30a4d8
2021-03-24 15:03:12.295583: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Loaded model in 0.014s.
Loading scorer from files deepspeech-0.9.3-models.scorer
Loaded scorer in 0.000314s.
Running inference.
why should one halt on the way
Inference took 1.329s for 2.735s audio file.
ds-tc-worker:test-github-actions ds-worker-admin$ deepspeech --model deepspeech-0.9.3-models.tflite --scorer deepspeech-0.9.3-models.scorer --audio audio/4507-16021-0012.wav --extended
Loading model from file deepspeech-0.9.3-models.tflite
TensorFlow: v2.3.0-6-g23ad988fcd
DeepSpeech: v0.10.0-alpha.3-57-g0b30a4d8
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2021-03-24 15:03:26.890383: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Data loss: Can't parse deepspeech-0.9.3-models.tflite as binary proto
Traceback (most recent call last):
  File "/usr/local/bin/deepspeech", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/deepspeech/client.py", line 119, in main
    ds = Model(args.model)
  File "/usr/local/lib/python3.9/site-packages/deepspeech/__init__.py", line 38, in __init__
    raise RuntimeError("CreateModel failed with '{}' (0x{:X})".format(deepspeech.impl.ErrorCodeToErrorMessage(status),status))
RuntimeError: CreateModel failed with 'Error reading the proto buffer model file.' (0x3005)
ds-tc-worker:test-github-actions ds-worker-admin$ pip3 install --upgrade deepspeech_tflite-0.10.0a3-cp39-cp39-macosx_10_9_x86_64.whl
Processing ./deepspeech_tflite-0.10.0a3-cp39-cp39-macosx_10_9_x86_64.whl
Requirement already satisfied: numpy in /usr/local/lib/python3.9/site-packages (from deepspeech-tflite==0.10.0a3) (1.20.1)
Installing collected packages: deepspeech-tflite
Successfully installed deepspeech-tflite-0.10.0a3
ds-tc-worker:test-github-actions ds-worker-admin$ deepspeech --model deepspeech-0.9.3-models.tflite --scorer deepspeech-0.9.3-models.scorer --audio audio/4507-16021-0012.wav --extended
Loading model from file deepspeech-0.9.3-models.tflite
TensorFlow: v2.3.0-6-g23ad988fcd
DeepSpeech: v0.10.0-alpha.3-57-g0b30a4d8
Loaded model in 0.0011s.
Loading scorer from files deepspeech-0.9.3-models.scorer
Loaded scorer in 0.0004s.
Running inference.
why should one halt on the way
Inference took 0.585s for 2.735s audio file.
ds-tc-worker:test-github-actions ds-worker-admin$
```

**Comment by lissyx** [comment]
@reuben Ready to merge on my side.

File: .github/workflows/macOS-amd64.yml
Comment (reuben) [inline .github/workflows/macOS-amd64.yml:168]: This is going to create `deepspeech-tf` packages which should be just `deepspeech` (or PyPI will reject upload).

**Review by reuben — APPROVED** [review]
Just one comment

File: .github/workflows/macOS-amd64.yml
Comment (lissyx) [reply .github/workflows/macOS-amd64.yml:168]: @reuben yes, but I don't know how to fix that properly :/

Comment (reuben) [reply .github/workflows/macOS-amd64.yml:168]: Pass build-flavor directly into the action and have a conditional there to set the correct project name?

**Pull request merged**
